File: fhgapp/src/main/java/de/jbamberger/fhgapp/ui/components/BaseFragment.java
package de.jbamberger.fhgapp.ui.components;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;

import javax.inject.Inject;

import de.jbamberger.fhgapp.di.Injectable;

/**
 * @author Jannik Bamberger (dev.jbamberger@gmail.com)
 */

public abstract class BaseFragment<T extends ViewModel> extends Fragment implements Injectable {

    @Inject
    ViewModelProvider.Factory viewModelFactory;

    private T viewModel;

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(getViewModelClass());
    }

    public abstract Class<T> getViewModelClass();
}

File: fhgapp/src/main/java/de/jbamberger/fhgapp/ui/components/BaseFragment.kt
package de.jbamberger.fhgapp.ui.components

import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import de.jbamberger.fhgapp.di.Injectable
import javax.inject.Inject

/**
 * @author Jannik Bamberger (dev.jbamberger@gmail.com)
 */

abstract class BaseFragment<T : ViewModel> : Fragment(), Injectable {

    @Inject
    internal lateinit var viewModelFactory: ViewModelProvider.Factory

    protected lateinit var viewModel: T

    abstract val viewModelClass: Class<T>

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel = ViewModelProvider(this, viewModelFactory).get(viewModelClass)
    }
}